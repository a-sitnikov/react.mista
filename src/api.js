//@flow
import fetchJsonp from 'fetch-jsonp'
import * as utils from './utils'

export const domain = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList    = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo     = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin         = localStorage.getItem('urlLogin') || 'ajax_login.php';
export const urlLogout        = localStorage.getItem('urlLogout') || 'users.php'; // users.php?&action=exit
export const urlCookies       = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections      = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage    = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id';
export const urlNewTopic      = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark   = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch        = localStorage.getItem('urlSearch') || 'ajax_find.php';

// Topics list
export type RequestTopicsList = {
    topics?:   number, // число тем, которые нужно вернуть (по умолчанию - 20);
    utime?:    number, // время (в формате unixtime) после которого надо вернуть ветки;
    callback?: string, // если указано, то ответ заворачивается в вызов функции, имя которой передано в параметре (реализация JSONP)
    forum?:    string, // раздел форума (1c|it|life);
    section_short_name?: string,
    user_id?:  string,
    mytopics?: string
}

export type ResponseTopicsListItem = {
    id: number,
    forum: string,
    sect1: string,
    sect2: string,
    v8: string,
    closed: number,
    down: number,
    text: string,
    utime: number,
    created: number,
    user: string,
    user0: string,
    is_voting: number,
    answ: number
}

export type ResponseTopicsList = Array<ResponseTopicsListItem>;

export const getTopicsList = async (params: RequestTopicsList): Promise<ResponseTopicsList> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicsList, params);
    return json;
} 


// Topics info
export type RequestInfo = {
    id: string // идентификатор (topic_id) темы
}

export type ResponseVoteItem = {
    select: string,
    result: number
}

export type ResponseInfo = {
    id: string,
    text: string,
    forum?: string,
    section?: string,
    created?: string,
    user_id?: string,
    user_name?: string,
    answers_count: string,
    down?: number,
    closed?: number,
    deleted?: number,
    is_voting?: number,
    voting?: Array<ResponseVoteItem>
}

export const defaultInfo = {
    id: "",
    text: "",
    answers_count: ""
}

export const getTopicInfo = async (params: RequestInfo): Promise<ResponseInfo> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicInfo, params);
    return json;
} 


// Topic messages
export type RequestMessages = {
    id:    number | string, // идентификатор (topic_id) темы
    from?: number, // с какого сообщения. если не указан, то с первого
    to?:   number  // до какого сообщения. если не указан, то from+10
}

export type ResponseMessage = {
    id:    string,
    n:     string,
    user:  string,
    userId: string,
    text:  string,
    utime: string,
    vote:  number
}

export type ResponseMessages = Array<ResponseMessage>;

export const getTopicMessages = async (params: RequestMessages): Promise<ResponseMessages> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicMessages, params);
    return json;
} 

export const getMessage = async (id: number | string, n: number): Promise<?ResponseMessage> =>  {
    const json = await getTopicMessages({
        id, 
        from: n, 
        to: n+1}
        );

    let data;
    if (json && json.length > 0)
        data = json.find(val => val.n === String(n));

    return data;
} 

export const getTopicMessagesCount = async (id: number | string): number => {
    const info = await getTopicInfo({id: String(id)});
    return +info.answers_count;
}

// Login
export type RequestLogin = {
    username:  string, // логин (ник) пользователя
    password:  string, // пароль пользователя
    callback?: string, // если указано, то ответ заворачивается в вызов функции, имя которой передано в параметре (реализация JSONP)
}

export type ResponseLogin = {
    userid: string,
    username: string,
    hashkey: string,
    error: string
}

export const getLogin = async (params: RequestLogin): Promise<ResponseLogin> =>  {
    const json = await fetchJsonpAndGetJson(urlLogin, params);
    return json;
} 

export const getLogout = async(): Promise<any> => {
    const params = { action: "exit" };
    await fetchJsonpAndGetJson(urlLogout, params);
}


// Cookies
export type ResponseCookies = {
    cookie: {
        entr_id: string,
        entr_key: string,
        entr_hash: string
    },
    session: {
        user_id: string,
        user_name: string,
        user_hash: string,
        is_moderator: number,
        light_moderator: number,
        section: string,
        show_moderator_tools: number,
        last_error?: string   
    }
}

export const getCookies = async (): Promise<ResponseCookies> =>  {
    const json = await fetchJsonpAndGetJson(urlCookies);
    return json;
} 


// Sections
export type ResponseSection = {
    id: number,
    forum: string,
    shortn: string,
    fulln: string,
    id: number
}

export type ResponseSections = Array<ResponseSection>;

export const getSections = async (): Promise<ResponseSections> =>  {
    const json = await fetchJsonpAndGetJson(urlSections);
    return json;
} 


// New message
export type RequestNewMessage = {
    message_text: string,
    action:       "new", 
    topic_id:     string,
    rnd:          number,
    voting_select?: number,
    as_admin?: boolean    
}

export const postNewMessage = async (params: RequestNewMessage): Promise<any> =>  {

    let url = urlNewMessage.replace(':id', params.topic_id);

    await fetch(`${domain}/${url}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
} 


// New topic
export type RequestNewTopic = {
    action:         "new", 
    rnd:            number,
    topic_text:     string,
    message_text:   string,
    target_section: string, 
    target_forum:   string,
    voting:         number,
    select1?:       string,    
    select2?:       string,    
    select3?:       string,    
    select4?:       string,    
    select5?:       string,    
    select6?:       string,    
    select7?:       string,    
    select8?:       string,    
    select9?:       string,    
    select10?:      string
}

export const postNewTopic = async (params: RequestNewTopic): Promise<any> =>  {

    await fetch(`${domain}/${urlNewTopic}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded'
        }    
    });
} 


//Bookmark
export type RequestBookmark = {
    id: string
}

export const postBookmark = async (params: RequestBookmark) => {

    await fetch(`${domain}/${urlAddBookmark}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
}


//Search
export type RequestSearch = {
    keywords: string
}

export type ResponseSearch = {

}

export const postSearch = async (params: RequestSearch) => {

    await fetch(`${domain}/${urlSearch}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
}

export const paramsToString = (paramsPrefix: string, params: ?{}): string => {

    if (!params)
        return '';

    let arr = [];
    for (let key in params) {
        if (params[key] !== undefined)
            arr.push(key + '=' + params[key]);
    }

    if (arr.length > 0)
        return paramsPrefix + arr.join('&');
    else    
        return '';
}

export const fetchJsonpAndGetJson = async (url: string, params: any): Promise<any> => {

    let fullUrl = `${domain}/${url}${paramsToString('?', params)}`; 
    const response =  await fetchJsonp(fullUrl);
    let responseJson = await response.json();
    let json;
    if (typeof(responseJson) === 'string') {
        
        try {
             json = JSON.parse(responseJson)    
        } catch(e) {
            json = utils.parseJSON(responseJson);
        }
    } else {
        json = responseJson;
    }
    return json;
}
