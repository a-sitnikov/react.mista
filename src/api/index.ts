export const domain: string = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList: string = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo: string = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages: string = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin: string = localStorage.getItem('urlLogin') || 'users.php?action=do_enter';
export const urlLogout: string = localStorage.getItem('urlLogout') || 'users.php?action=exit';
export const urlCookies: string = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections: string = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage: string = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id&page=1';
export const urlNewTopic: string = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark: string = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch: string = localStorage.getItem('urlSearch') || 'ajax_find.php';

export * from './cookies'
export * from './login'
export * from './new_message'
export * from './newtopic'
export * from './sections'
export * from './topic_info'
export * from './topic_messages'
export * from './topics_list'
