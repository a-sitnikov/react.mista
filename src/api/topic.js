import { domain, urlTopicInfo, urlTopicMessages, fetchAndGetJson, fetchJsonpAndGetJson } from '.'

// Topics info
export type RequestInfo = {
  id: string // идентификатор (topic_id) темы
}

export type ResponseVoteItem = {
  select: string,
  result: number
}

export type ResponseInfo = {
  id: string,
  text: string,
  forum?: string,
  section?: string,
  created?: string,
  user_id?: string,
  user_name?: string,
  answers_count: string,
  down?: number,
  closed?: number,
  deleted?: number,
  is_voting?: number,
  voting?: Array<ResponseVoteItem>
}

export const getTopicInfo = async (params: RequestInfo): Promise<ResponseInfo> => {
  const json = await fetchAndGetJson(urlTopicInfo, params);
  return json;
}


// Topic messages
export type RequestMessages = {
  id: number | string, // идентификатор (topic_id) темы
  from?: number, // с какого сообщения. если не указан, то с первого
  to?: number  // до какого сообщения. если не указан, то from+10
}

export type ResponseMessage = {
  id: string,
  n: string,
  user: string,
  userId: string,
  text: string,
  utime: string,
  vote: number
}

export type ResponseMessages = Array<ResponseMessage>;

export const getTopicMessages = async (params: RequestMessages): Promise<ResponseMessages> => {
  const json = await fetchAndGetJson(urlTopicMessages, params);
  return json;
}

export const getMessage = async (id: number | string, n: number): Promise<?ResponseMessage> => {
  
  let from = n;
  let to = +n;
  if (n === 0) //does't work from=0&to=0
    to++;

  const json = await getTopicMessages({
    id,
    from,
    to
  }
  );

  let data;
  if (json && json.length > 0)
    data = json.find(val => val.n === String(n));

  return data;
}

export const getTopicMessagesCount = async (id: number | string): Promise<number> => {
  const info = await getTopicInfo({ id: String(id) });
  return +info.answers_count;
}