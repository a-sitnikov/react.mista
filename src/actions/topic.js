//@flow
import * as API from 'src/api'
import type { Dispatch } from 'redux'

import type { ResponseInfo, ResponseMessage, ResponseMessages } from 'src/api'

export type REQUEST_TOPIC = {
    type: 'REQUEST_TOPIC'
}

export type RECEIVE_TOPIC = {
    type: 'RECEIVE_TOPIC',
    info: ResponseInfo, 
    item0: ResponseMessage, 
    items: ResponseMessages,
    receivedAt: Date
}

export type RECEIVE_TOPIC_FAILED = {
    type: 'RECEIVE_TOPIC_FAILED',
    error: any,
    receivedAt: Date
}

export type TopicAction = REQUEST_TOPIC | RECEIVE_TOPIC | RECEIVE_TOPIC_FAILED;

export const requestTopic = () => ({
    type: 'REQUEST_TOPIC'
})

export const receiveTopic = (info: ResponseInfo, item0: ?ResponseMessage, items: ResponseMessages) => {

    return {
        type: 'RECEIVE_TOPIC',
        info,
        item0,
        items,
        receivedAt: Date.now()
    }
}


export const fetchTopic = (params: any, item0: ?ResponseMessage) => async (dispatch: Dispatch<*>) => {

    dispatch(requestTopic())

    let info;
    try {
        info = await API.getTopicInfo({id: params.id});
    } catch(e) {
        console.error(e);
        info = {
            id: params.id,
            text: '',
            answers_count: "0"
        };
    }   

    try {    
        let page = params.page || 1;

        let _item0 = item0;
        let _items;
        if (page === 'last20') {

            if (+info.answers_count > 21) {

                if (!_item0) {
                    let items = await API.getTopicMessages({
                        id: params.id,
                        from: 0,
                        to: 1
                    }); 
                    _item0 = items[0];
                }    

                let first = +info.answers_count - 20;
                _items = await API.getTopicMessages({
                    id: params.id,
                    from: first,
                    to: 1010
                }); 

            } else {
                let items = await API.getTopicMessages({
                    id: params.id,
                    from: 0,
                    to: 1010
                });               
                _item0 = items[0];
                _items = items.slice(1);
            }

        } else {

            page = +page;
            let first = 0;
            let last = page * 100 - 1;

            if (page > 1) {

                first = (page - 1) * 100 + 1;
                if (!_item0) {
                    let items = await API.getTopicMessages({
                        id: params.id,
                        from: 0,
                        to: 1
                    });
                    _item0 = items[0];
                }

                _items = await API.getTopicMessages({
                    id: params.id,
                    from: first,
                    to: last
                });                

            } else {
                if (_item0)
                    first = 1;
                else
                    first = 0;

                let items = await API.getTopicMessages({
                    id: params.id,
                    from: first,
                    to: last
                });  

                if (_item0) {
                    _items = items;
                } else {
                    _item0 = items[0];
                    _items = items.slice(1);
                }
            }
        
        }
    
        if (info.answers_count === "0" && _items.length > 0)
            info.answers_count = _items[_items.length - 1].n;

        dispatch(receiveTopic(info, _item0, _items));

    } catch (error) {

        console.error('Failed to fetch topic:', error);

        dispatch({
            type: 'RECEIVE_TOPIC_FAILED',
            error,
            receivedAt: Date.now()
        });

    }
}

const shouldFetch = (state) => {
    const topic = state.topic;
    if (!topic) {
        return true
    }
    if (topic.isFetching) {
        return false
    }
    return true
}

export const fetchTopicIfNeeded = (params: any, item0: ?ResponseMessage) => (dispatch: Dispatch<*>, getState: any) => {
    if (shouldFetch(getState())) {
        return dispatch(fetchTopic(params, item0));
    }
}

export type FetchNewMessageseParams = {
    id: number | string,
    last: number    
}

export const fetchNewMessages = (params: FetchNewMessageseParams) => async (dispatch: Dispatch<*>) => {

    dispatch({
        type: 'REQUEST_NEW_MESSAGES'
    });

    try {
        const json = await API.getTopicMessages({
            id: params.id,
            from: params.last + 1,
            to: 1002
        });

        dispatch({
            type: 'RECEIVE_NEW_MESSAGES',
            items: json,
            receivedAt: Date.now()
        });

    } catch (error) {
        console.error('Failed to fetch new messages:', error);

        dispatch({
            type: 'RECEIVE_NEW_MESSAGES_FAILED',
            error,
            receivedAt: Date.now()
        });
    }

}

export const fetchNewMessagesIfNeeded = (params: FetchNewMessageseParams) => (dispatch: Dispatch<*>, getState: any) => {
    if (shouldFetch(getState())) {
        return dispatch(fetchNewMessages(params));
    }
}