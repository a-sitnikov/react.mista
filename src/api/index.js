//@flow
import fetchJsonp from 'fetch-jsonp'
import * as utils from 'src/utils'

export const domain: string = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList: string = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo: string = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages: string = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin: string = localStorage.getItem('urlLogin') || 'users.php?action=do_enter';
export const urlLogout: string = localStorage.getItem('urlLogout') || 'users.php?action=exit';
export const urlCookies: string = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections: string = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage: string = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id&page=1';
export const urlNewTopic: string = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark: string = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch: string = localStorage.getItem('urlSearch') || 'ajax_find.php';

// New message
export type RequestNewMessage = {
  message_text: string,
  action: "new",
  topic_id: string,
  rnd: number,
  voting_select?: number,
  as_admin?: boolean
}

export const postNewMessage = async (params: RequestNewMessage): Promise<any> => {

  let url = urlNewMessage.replace(':id', params.topic_id);

  let result = await fetch(`${domain}/${url}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
    headers: {
      'Accept': 'text/html,application/xhtml+xml,application/xml',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    redirect: 'follow'
  });

  return result;
}


// New topic
export type RequestNewTopic = {
  action: "new",
  rnd: number,
  topic_text: string,
  message_text: string,
  target_section: string,
  target_forum: string,
  voting: number,
  select1?: string,
  select2?: string,
  select3?: string,
  select4?: string,
  select5?: string,
  select6?: string,
  select7?: string,
  select8?: string,
  select9?: string,
  select10?: string
}

export const postNewTopic = async (params: RequestNewTopic): Promise<any> => {

  await fetch(`${domain}/${urlNewTopic}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    }
  });
}


//Bookmark
export type RequestBookmark = {
  id: string
}

export const postBookmark = async (params: RequestBookmark) => {

  await fetch(`${domain}/${urlAddBookmark}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
  });
}

export const postSearch = async (params: RequestSearch) => {

  await fetch(`${domain}/${urlSearch}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
  });
}

export const paramsToString = (paramsPrefix: string, params: ?{}): string => {

  if (!params)
    return '';

  let arr = [];
  for (let key in params) {
    if (params[key] !== undefined)
      arr.push(key + '=' + params[key]);
  }

  if (arr.length > 0)
    return paramsPrefix + arr.join('&');
  else
    return '';
}

export const fetchJsonpAndGetJson = async (url: string, params: any): Promise<any> => {

  let fullUrl = `${domain}/${url}${paramsToString('?', params)}`;
  const response = await fetchJsonp(fullUrl);
  let responseJson = await response.json();
  let json;
  if (typeof (responseJson) === 'string') {

    try {
      json = JSON.parse(responseJson)
    } catch (e) {
      json = utils.parseJSON(responseJson);
    }
  } else {
    json = responseJson;
  }
  return json;
}

export const fetchAndGetJson = async (url: string, params: any, options: any): Promise<any> => {

  let fullUrl = `${domain}/${url}${paramsToString('?', params)}`;
  const response = await fetch(fullUrl, options);
  let responseJson = await response.json();
  let json;
  if (typeof (responseJson) === 'string') {

    try {
      json = JSON.parse(responseJson)
    } catch (e) {
      json = utils.parseJSON(responseJson);
    }
  } else {
    json = responseJson;
  }
  return json;
}