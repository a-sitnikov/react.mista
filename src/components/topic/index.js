//@flow
import React, { Component } from 'react'
import { connect } from 'react-redux'
import queryString from 'query-string'
import { fetchTopicIfNeeded, fetchNewMessagesIfNeeded } from 'src/actions/topic'

import type { DefaultProps, Location } from 'src/components'
import type { ResponseInfo, ResponseMessage, ResponseMessages } from 'src/api'
import type { State } from 'src/reducers'
import { defaultTopicState } from 'src/reducers/topic'
import { OptionsItems } from 'src/reducers/options'

import Header from './header'
import TopicInfo from './topic_info'
import Row from './row'
import Footer from './footer'
import NewMessage from './new_message';
import { getMaxPage } from 'src/utils';

type TopicLocationParams = {
    id: string,
    page?: string,
}

type Column = {
    name: string,
    width?: string
}

type TopicProps = {
    info: ResponseInfo,
    item0?: ResponseMessage,
    items: ResponseMessages,
    error?: any,
    options: OptionsItems
}

type Props = {
    fetchTopicIfNeeded: any,
    fetchNewMessagesIfNeeded: any
} & DefaultProps & TopicProps

class Topic extends Component<Props> {

    onPostNewMessageSuccess;
    updateTopic;
    autoUpdate;
    locationParams: TopicLocationParams;
    location: Location;
    columns: Array<Column>;
    timer;

    constructor(props) {
        super(props);
        this.onPostNewMessageSuccess = this.onPostNewMessageSuccess.bind(this);
        this.updateTopic = this.updateTopic.bind(this)
        this.autoUpdate = this.autoUpdate.bind(this)
        
        this.locationParams = {id: ''};
        
        this.columns = [
            { name: 'Автор', width: '165px' },
            { name: 'Текст' }
        ];
        
    }

    componentDidMount() {
        
        this.location = this.props.location;
        this.updateTopic();

        if (this.props.options.autoRefreshTopic === 'true') {
            
            let autoRefreshTopicInterval = +this.props.options.autoRefreshTopicInterval;
            if (autoRefreshTopicInterval < 60) autoRefreshTopicInterval = 60;

            this.timer = setInterval(this.autoUpdate, autoRefreshTopicInterval * 1000);
        }
        
    }

    componentWillUnmount() {
        clearInterval(this.timer);
    }

    componentWillReceiveProps(props: Props) {
        
        if (props.info.text && document.title !== props.info.text) {
            document.title = props.info.text;
        }

        if (this.location.search !== props.location.search) {
            this.location = props.location;
            this.updateTopic();
        }
    }
    
    autoUpdate() {
        const { info, fetchNewMessagesIfNeeded } = this.props;
        const isLastPage = (this.locationParams.page === 'last20' || this.locationParams.page === getMaxPage(parseInt(info.answers_count, 10)));

        if (isLastPage) 
            fetchNewMessagesIfNeeded({
                id: info.id,
                last: parseInt(info.answers_count, 10)
            })
    }

    updateTopic() {
       
        let { fetchTopicIfNeeded, item0 } = this.props;
        let locationParams = queryString.parse(this.location.search);
        locationParams.hash = this.location.hash;

        if (!locationParams.page)
            locationParams.page = 1;

        else if (locationParams.page !== 'last20')  {
            locationParams.page = parseInt(locationParams.page, 10);
            if (isNaN(locationParams.page))
                locationParams.page = 1;
        }  

        if (locationParams.id !== this.locationParams.id)
            item0 = null;

        this.locationParams = locationParams;
        fetchTopicIfNeeded(this.locationParams, item0);
    }

    onPostNewMessageSuccess() {

        const { fetchNewMessagesIfNeeded, info } = this.props;

        const isLastPage = (this.locationParams.page === 'last20' || this.locationParams.page === getMaxPage(parseInt(info.answers_count, 10)));

        if (isLastPage)
            fetchNewMessagesIfNeeded({
                id: info.id,
                last: parseInt(info.answers_count, 10)
            });

    }

    render() {
        const { items, item0, error } = this.props;

        let errorElem;
        if (error)
            errorElem = (
                <div >
                    <p className="error">ОШИБКА</p>
                    <p className="error">{error.message}</p>
                </div>
            )

        return (
            <div  >
                {errorElem}
                <Header currentPage={this.locationParams.page} />
                <table id='tm' className="border2" style={{ width: "100%", margin: "10px auto 0px auto" }}>
                    <colgroup>
                        {this.columns.map((item, i) => (
                            <col key={i} style={{ width: item.width }} />
                        ))}
                    </colgroup>
                    <tbody>
                        <TopicInfo />
                        <Row key='0' columns={this.columns} data={item0}/>
                        {items.map((item, i) => (
                            <Row key={item.n} columns={this.columns} data={item}/>
                        ))}
                    </tbody>
                </table>
                <Footer params={this.locationParams} />
                <NewMessage onPostSuccess={this.onPostNewMessageSuccess} />
            </div>
        )
    }
}

const mapStateToProps = (state: State): TopicProps => {

    const {
        isFetching,
        lastUpdated,
        info,
        item0,
        items,
        error
    } = state.topic || defaultTopicState;

    return {
        login: state.login,
        info,
        item0,
        items,
        isFetching,
        lastUpdated,
        error,
        options: state.options.items
    }
}

const mapDispatchToProps = (dispatch) => ({
    fetchTopicIfNeeded: (...params) => dispatch(fetchTopicIfNeeded(...params)),
    fetchNewMessagesIfNeeded: (...params) => dispatch(fetchNewMessagesIfNeeded(...params))   
})

export default connect(mapStateToProps, mapDispatchToProps)(Topic);