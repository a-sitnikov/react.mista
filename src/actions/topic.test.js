import { fetchTopic } from './topic'
import * as API from '../api'

const item0 = {
    id: "0", n: "0", text: "Пост 0"
}

describe('Topic.Actions', ()=>{

    beforeEach(() => {

        API.getTopicInfo = jest.fn(params => {
            return {
                id: "1",
                text: "Тестовая ветка",
                answers_count: "50"
            }
        })

        API.getTopicMessages = jest.fn(params => {

            return [
                {
                    id: "1",
                    n: params.from,
                    text: "Пост " + params.from
                },
                {
                    id: "1",
                    n: +params.from + 1,
                    text: "Пост " + (+params.from + 1)
                },
            ]
        })

        Date.now = jest.fn(() => 1);
    })

    it('+++ page 1, no item 0', async () => {

        const dispatch = jest.fn();
        await fetchTopic({
            id: "1",
            page: "1"
        }, null)(dispatch);
        
        expect(dispatch.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[1][0]).toMatchSnapshot();

    })
    
    it('+++ page 1, has item 0', async () => {
        
        const dispatch = jest.fn();
        await fetchTopic({
            id: "1",
            page: "1"
        }, item0)(dispatch)
        
        expect(dispatch.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[1][0]).toMatchSnapshot();
    })

    it('+++ page 2, no item 0', async () => {
        const dispatch = jest.fn();
        await fetchTopic({
            id: "1",
            page: "2"
        }, null)(dispatch)
        
        expect(dispatch.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[1][0]).toMatchSnapshot();
    });
    
    it('+++ page 2, has item 0', async () => {
        const dispatch = jest.fn();
        await fetchTopic({
            id: "1",
            page: "2"
        }, item0)(dispatch)
        
        expect(dispatch.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[1][0]).toMatchSnapshot();

    });
    
    it('+++ page last20, no item 0', async () => {
        const dispatch = jest.fn();
        await fetchTopic({
            id: "1",
            page: "last20"
        }, null)(dispatch)
        
        expect(dispatch.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[1][0]).toMatchSnapshot();

    });
    
    it('+++ page last20, less than 20 answers, no item 0', async () => {
        API.getTopicInfo = jest.fn(params => {
            return {
                id: "1",
                text: "Тестовая ветка",
                answers_count: "18"
            }
        })
        
        const dispatch = jest.fn();
        await fetchTopic({
            id: "1",
            page: "last20"
        }, null)(dispatch)
        
        expect(dispatch.mock.calls.length).toBe(2);
        expect(dispatch.mock.calls[1][0]).toMatchSnapshot();
    });
})