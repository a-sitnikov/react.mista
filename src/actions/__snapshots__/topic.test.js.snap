// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic.Actions +++ page 1, has item 0 1`] = `
Object {
  "info": Object {
    "answers_count": "50",
    "id": "1",
    "text": "Тестовая ветка",
  },
  "item0": Object {
    "id": "0",
    "n": "0",
    "text": "Пост 0",
  },
  "items": Array [
    Object {
      "id": "1",
      "n": 1,
      "text": "Пост 1",
    },
    Object {
      "id": "1",
      "n": 2,
      "text": "Пост 2",
    },
  ],
  "receivedAt": 1,
  "type": "RECEIVE_TOPIC",
}
`;

exports[`Topic.Actions +++ page 1, no item 0 1`] = `
Object {
  "info": Object {
    "answers_count": "50",
    "id": "1",
    "text": "Тестовая ветка",
  },
  "item0": Object {
    "id": "1",
    "n": 0,
    "text": "Пост 0",
  },
  "items": Array [
    Object {
      "id": "1",
      "n": 1,
      "text": "Пост 1",
    },
  ],
  "receivedAt": 1,
  "type": "RECEIVE_TOPIC",
}
`;

exports[`Topic.Actions +++ page 2, has item 0 1`] = `
Object {
  "info": Object {
    "answers_count": "50",
    "id": "1",
    "text": "Тестовая ветка",
  },
  "item0": Object {
    "id": "0",
    "n": "0",
    "text": "Пост 0",
  },
  "items": Array [
    Object {
      "id": "1",
      "n": 100,
      "text": "Пост 100",
    },
    Object {
      "id": "1",
      "n": 101,
      "text": "Пост 101",
    },
  ],
  "receivedAt": 1,
  "type": "RECEIVE_TOPIC",
}
`;

exports[`Topic.Actions +++ page 2, no item 0 1`] = `
Object {
  "info": Object {
    "answers_count": "50",
    "id": "1",
    "text": "Тестовая ветка",
  },
  "item0": Object {
    "id": "1",
    "n": 0,
    "text": "Пост 0",
  },
  "items": Array [
    Object {
      "id": "1",
      "n": 100,
      "text": "Пост 100",
    },
    Object {
      "id": "1",
      "n": 101,
      "text": "Пост 101",
    },
  ],
  "receivedAt": 1,
  "type": "RECEIVE_TOPIC",
}
`;

exports[`Topic.Actions +++ page last20, less than 20 answers, no item 0 1`] = `
Object {
  "info": Object {
    "answers_count": "18",
    "id": "1",
    "text": "Тестовая ветка",
  },
  "item0": Object {
    "id": "1",
    "n": 0,
    "text": "Пост 0",
  },
  "items": Array [
    Object {
      "id": "1",
      "n": 1,
      "text": "Пост 1",
    },
  ],
  "receivedAt": 1,
  "type": "RECEIVE_TOPIC",
}
`;

exports[`Topic.Actions +++ page last20, no item 0 1`] = `
Object {
  "info": Object {
    "answers_count": "50",
    "id": "1",
    "text": "Тестовая ветка",
  },
  "item0": Object {
    "id": "1",
    "n": 0,
    "text": "Пост 0",
  },
  "items": Array [
    Object {
      "id": "1",
      "n": 30,
      "text": "Пост 30",
    },
    Object {
      "id": "1",
      "n": 31,
      "text": "Пост 31",
    },
  ],
  "receivedAt": 1,
  "type": "RECEIVE_TOPIC",
}
`;
